File: fixtures/ruby/spec/calculator_spec.rb
# frozen_string_literal: true

require_relative "../lib/calculator"
require_relative "../lib/errors"

RSpec.describe Calculator do
  subject(:calculator) { described_class.new }

  describe "#add" do
    it "returns the sum of two numbers" do
      expect(calculator.add(2, 3)).to eq(5)
    end
  end

  describe "#subtract" do
    it "returns the difference of two numbers" do
      expect(calculator.subtract(5, 3)).to eq(2)
    end
  end

  describe "#multiply" do
    it "returns the product of two numbers" do
      expect(calculator.multiply(4, 3)).to eq(12)
    end
  end

  describe "#divide" do
    it "returns the quotient of two numbers" do
      expect(calculator.divide(10, 2)).to eq(5.0)
    end

    it "raises DivisionByZeroError when dividing by zero" do
      expect { calculator.divide(1, 0) }.to raise_error(DivisionByZeroError)
    end
  end

  describe "#history" do
    it "tracks calculation results" do
      calculator.add(1, 2)
      calculator.multiply(3, 4)
      expect(calculator.history).to eq([3, 12])
    end
  end

  describe "#clear_history" do
    it "clears the history" do
      calculator.add(1, 2)
      calculator.clear_history
      expect(calculator.history).to eq([])
    end
  end
end

File: fixtures/lua/spec/calculator_spec.lua
local Calculator = require("src.calculator")

describe("Calculator", function()
    local calc

    before_each(function()
        calc = Calculator.new()
    end)

    it("adds two numbers", function()
        assert.are.equal(5, calc:add(2, 3))
    end)

    it("subtracts two numbers", function()
        assert.are.equal(2, calc:subtract(5, 3))
    end)

    it("multiplies two numbers", function()
        assert.are.equal(12, calc:multiply(4, 3))
    end)

    it("divides two numbers", function()
        assert.are.equal(5, calc:divide(10, 2))
    end)

    it("raises error on division by zero", function()
        assert.has_error(function()
            calc:divide(1, 0)
        end)
    end)

    it("tracks history", function()
        calc:add(1, 2)
        calc:multiply(3, 4)
        local history = calc:get_history()
        assert.are.equal(3, history[1])
        assert.are.equal(12, history[2])
    end)

    it("clears history", function()
        calc:add(1, 2)
        calc:clear_history()
        assert.are.same({}, calc:get_history())
    end)
end)
